flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBWriteBatchWrapper.java

```diff
@@ -113,4 +126,16 @@ public void close() throws RocksDBException {
 		}
 		IOUtils.closeQuietly(batch);
 	}
+
+	private void flushIfNeeded() throws RocksDBException {
+		boolean needFlush = batch.count() == capacity || (batchSize > 0 && batch.getDataSize() >= batchSize);
+		if (needFlush) {
+			flush();
+		}
+	}
+
+	@VisibleForTesting
+	long getDataSize() {
+		return batch.getDataSize();
```

Comment:
Do we know how expensive this operation is (JNI call overhead, etc.)?
If it might be expensive, it should be very simply to just keep a byte count in the wrapper directly.

Reply:
I run some microbenchmark locally, it will consume ~ 40 us per call, I've adjusted the implementation.
```
Benchmark                  Mode  Cnt   Score    Error  Units
BBenchmark.testWriteBatch    ss   10  44.186 ± 14.880  us/op
```

Reply:
A little bit curious about how much the new implementation (additional calculation on data size) could improve the performance. Mind adding some comparison between the old (jni call) and new implementation here through micro benchmark? @klion26  Thanks.

Reply:
Yesterday, I used singleshot mode to benchmark this, now I ran benchmark both singleshot and averagetime, `averagetime` mode will get ~ 2% improvment

```
Benchmark                                     Mode  Cnt  Score   Error  Units
BBenchmark.testGetDataSizeAfterPut            avgt   15  0.369 ± 0.067  us/op
BBenchmark.testRawGetDataSizeAfterPut         avgt   15  0.389 ± 0.088  us/op
BBenchmark.testGetDataSizeAfterPutOneShot       ss   15  2.458 ± 1.467  us/op
BBenchmark.testRawGetDataSizeAfterPutOneShot    ss   15  2.024 ± 0.223  us/op
```

attached the benchmark code below
```
@Benchmark
@BenchmarkMode(Mode.SingleShotTime)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
public long testGetDataSizeAfterPutOneShot() throws RocksDBException {
	writeBatch.put(handle, dummy, dummy);
	return  2 + calculateVarint32Length(dummy.length) + dummy.length + calculateVarint32Length(dummy.length) + dummy.length;
}

private int calculateVarint32Length(int input) {
	if (input < (1 << 7)) {
		return 1;
	}
	if (input < (1 << 14)) {
		return 2;
	}
	if (input < (1 << 21)) {
		return 3;
	}
	if (input < (1 << 28)) {
		return 4;
	}
	return 5;
}

@Benchmark
@BenchmarkMode(Mode.SingleShotTime)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
public long testRawGetDataSizeAfterPutOneShot() throws RocksDBException {
	writeBatch.put(handle, dummy, dummy);
	return writeBatch.getDataSize();
}

@Benchmark
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
public long testGetDataSizeAfterPut() throws RocksDBException {
      writeBatch.put(handle, dummy, dummy);
      return  2 + calculateVarint32Length(dummy.length) + dummy.length + calculateVarint32Length(dummy.length) + dummy.length;
}

@Benchmark
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
public long testRawGetDataSizeAfterPut() throws RocksDBException {
       writeBatch.put(handle, dummy, dummy);
        return writeBatch.getDataSize();
 }
```

Reply:
IMO the `AverageTime` mode reflects the real world usage of `RocksDBWriteBatchWrapper` better than `SingleShot`, and normally we see 2% as a normal disturbance. With these perf data I'm biased on using the JNI call directly. Wdyt? @StephanEwen @klion26 Thanks.

Reply:
Using our own implemented `getDataSize()` we can gain 2% performance but need to track the format envolving in RocksDB, on the other hand, using JNI call directly we don't need to track the format envolving in RocksDB, 
As the performance gained is not much big, so I'm biased on using the JNI call directly also.

Reply:
Sounds fine to just call rocksDB directly, rather than calculating this ourselves.
I woudl also be fine with a more approximate calculation, where we always assume 2 or 4 bytes for the length. But this seems minor.